drivers: mbox: Add support for TI OMAP mailbox

vaishnavachath opened the pull request:
TI OMAP mailbox is the inter-processor mailbox IP found in TI K3 devices (AM62X, AM64X, J721E .etc). The mailbox hardware uses  a queued mailbox interrupt mechanism that provides a communication channel between processors through a set of registers and their associated interrupt signals by sending and receiving messages. The interrupt/bank associated with each processor entity is found through the  usr_id property from device tree.

The bindings and implementation are kept similar to the implementation in Linux, driver implementation logic is similar to other drivers in mbox subsystem.

Tested on AM62X SK for MBOX ping and MBOX Data sample between A53 and M4 running Zephyr : 
https://pastebin.com/f88f7NQJ

vaishnavachath committed:
drivers: mbox: Add support for TI OMAP mailbox

TI OMAP mailbox is the inter-processor mailbox IP found in TI
K3 devices (AM62X, AM64X, J721E .etc). The mailbox hardware uses
a queued mailbox interrupt mechanism that provides a communication
channel between processors through a set of registers and their
associated interrupt signals by sending and receiving messages.
The interrupt/bank associated with each processor entity is found
through the  usr_id property from device tree.

Signed-off-by: Vaishnav Achath <vaishnav.a@ti.com>

vaishnavachath committed:
dts: arm: ti: am62x_m4: Add mailbox nodes

Add TI OMAP interprocessor mailbox nodes for AM62X M4,
the user ID assignment is as per thec corresponding mailbox
interrupt assignment for the core.

More details can be found in the device TRM Mailbox section:
https://www.ti.com/lit/ug/spruiv7a/spruiv7a.pdf

Signed-off-by: Vaishnav Achath <vaishnav.a@ti.com>

vaishnavachath committed:
dts: arm64: ti: am62x_a53: Add mailbox nodes

Add TI OMAP interprocessor mailbox nodes for AM62X A53,
the user ID assignment is as per thec corresponding mailbox
interrupt assignment for the core. Also while at it update the
supported feature list in corresponding boards.

More details can be found in the device TRM Mailbox section:
https://www.ti.com/lit/ug/spruiv7a/spruiv7a.pdf

Signed-off-by: Vaishnav Achath <vaishnav.a@ti.com>

dnltz commented on dts/arm64/ti/ti_am62x_a53.dtsi:
```diff
@@ -133,4 +133,13 @@
 		status = "disabled";
 	};
+
+	mbox:mailbox0@29000000 {
```
```suggestion
	mbox0: mailbox0@29000000 {
```

dnltz commented on dts/arm/ti/am62x_m4.dtsi:
```diff
@@ -39,6 +39,15 @@
 		#clock-cells = <0>;
 	};
 
+	mbox:mailbox0@29000000 {
```
```suggestion
	mbox0: mailbox0@29000000 {
```

dnltz commented on dts/bindings/mbox/ti,omap-mailbox.yaml:
```diff
@@ -0,0 +1,23 @@
+  interrupts:
+    required: true
+
+  usr_id:
```
I think dashes are more often used in device trees as separator like `interrupt-cells`
```suggestion
  usr-id:
```

dnltz commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+				  CONFIG_MBOX_INIT_PRIORITY,	\
+			      &omap_mailbox_driver_api)
+
+#define MAILBOX_INST(idx) MAILBOX_INSTANCE_DEFINE(idx);
```
Looks like you could drop this and use `MAILBOX_INSTANCE_DEFINE` directly.

vaishnavachath replied:
Done

dnltz commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+		irq_enable(DT_INST_IRQN(idx));	\
+		return 0;	\
+	}	 \
+	DEVICE_DT_INST_DEFINE(idx, omap_mailbox_##idx##_init, \
+				  NULL, &omap_mailbox_##idx##_data,	\
+			      &omap_mailbox_##idx##_config, POST_KERNEL, \
+				  CONFIG_MBOX_INIT_PRIORITY,	\
+			      &omap_mailbox_driver_api)
```
Those lines are not indendet at the same level. More a nitpick :)

dnltz commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+	const static struct omap_mailbox_config omap_mailbox_##idx##_config = {	\
+		.usr_id =  DT_INST_PROP(idx, usr_id),	\
+	};	\
+	static int omap_mailbox_##idx##_init(const struct device *dev)	\
```
you could save the idx number in the `omap_mailbox_##idx##_config` and call a generic init function. Saves some code size when multiple mailbox drivers are enabled... your approach works too.

vaishnavachath replied:
Will fix this, this is much better as few other devices have multiple (10+) mailbox instances.

vaishnavachath replied:
I have skipped this for now, if it is okay I will fix it later when adding multiple instances

dnltz commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+	uint32_t irqstatus;
+
+	regs = (struct omap_mailbox_regs *)data->mapped_base;
+	if (channel >= MAILBOX_MAX_CHANNELS) {
+		return -EINVAL;
+	}
```
No need for brackets here.
```suggestion
	if (channel >= MAILBOX_MAX_CHANNELS)
		return -EINVAL;
```

dnltz commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+
+#define MAILBOX_IRQ_NEWMSG(m)		(1 << (2 * (m)))
+#define MAILBOX_IRQ_NOTFULL(m)		(1 << (2 * (m) + 1))
+#define OMAP_MAILBOX_NUM_MSGS 16
+#define MAILBOX_MAX_CHANNELS 16
+#define OMAP_MAILBOX_NUM_USERS 4
+#define MAILBOX_MBOX_SIZE    sizeof(uint32_t)
```
You could indent the values at the same level.

dnltz commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+	struct omap_mailbox_data *data = dev->data;
+	const struct omap_mailbox_config *cfg = dev->config;
+
+	volatile struct omap_mailbox_regs *regs = (struct omap_mailbox_regs *)data->mapped_base;
```
You could consider using defines to access driver related data like in https://github.com/zephyrproject-rtos/zephyr/blob/9f0cd47c784e4d2099f59899d17a6ba5f3d4be55/drivers/gpio/gpio_davinci.c#L26

dnltz commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+		return 0;
+	}
+
+	if (msg->size != MAILBOX_MBOX_SIZE) {
```
Why not per-initialize `data32` with zero and change this check to `<= MAILBOX_MBOX_SIZE`?

dnltz reviewed (commented):
Some feedback but looks good. I have one question. Why did you call this driver `omap` while we called the gpio driver `davinci`? What's the difference and what would be the reason to call a driver `davanci` or `omap`? As an external, non-TI engineer, I think it would make more sense to call all drivers identical to make it easier to identify them. But maybe you have a good explanation to choose `omap`.

vaishnavachath commented:
> Some feedback but looks good. I have one question. Why did you call this driver `omap` while we called the gpio driver `davinci`? What's the difference and what would be the reason to call a driver `davanci` or `omap`? As an external, non-TI engineer, I think it would make more sense to call all drivers identical to make it easier to identify them. But maybe you have a good explanation to choose `omap`.

Thank you for the feedback @dnltz I will work on your review comments, the reason behind naming these drivers with these legacy names is to match the driver names and compatibles in Linux kernel, for example the mailbox driver is here : https://github.com/torvalds/linux/blob/master/drivers/mailbox/omap-mailbox.c , the idea is to make some reuse from the upstream Linux offering, so that adding future nodes in Zephyr becomes easier and the Linux device tree can be taken as reference, an exact 1:1 match is not always possible but the idea is to keep the differences in compatible/bindings .etc same to enable reuse, same was the case for gpio-davinci driver.

aleeraser commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,194 @@
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
+	if (!msg) {
+		if (!regs->fifo_status[channel])
+			regs->message[channel] = 0;
+
+		return 0;
+	}
```
As far as I understood, this checks whether the FIFO is full. I think that in this case this API is [supposed to return](https://docs.zephyrproject.org/latest/hardware/peripherals/mbox.html#c.mbox_send) `-EBUSY`.

Something like:

```suggestion
	if (regs->fifo_status[channel]) {
		return -EBUSY;
	}

	if (!msg) {
		regs->message[channel] = 0;
		return 0;
	}
```

aleeraser commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,194 @@
+
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
```
This API is [supposed to return](https://docs.zephyrproject.org/latest/hardware/peripherals/mbox.html#c.mbox_set_enabled) `-EALREADY` if the channel is already enabled.

Since you're anyway tracking enabled channel via `channel_enable`:

```suggestion

	if (data->enabled_channels[channel]) {
		return -EALREADY;
	}

```

gramsay0 commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+	{	\
+		struct omap_mailbox_data *data = dev->data;	\
+		mem_addr_t addr;	\
+		device_map((mm_reg_t *)&addr, DT_INST_REG_ADDR(idx),	\
```
I think the `DEVICE_MMIO_*` macros are preferred for this

vaishnavachath replied:
Thanks for the review @gramsay0 , will fix this and other comments.

vaishnavachath replied:
Done, added to TI K3 MAINTAINERS entry

gramsay0 commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+
+	volatile struct omap_mailbox_regs *regs = DEV_BASE(data);
+	volatile struct omap_mailbox_irq_regs *irq_regs = &regs->irq_regs[cfg->usr_id];
+	uint32_t flags = irq_regs->status_clear;
```
- Is the IRQ status register the same as the IRQ clear register?
- Is the mailbox FIFO based?
If so, there should probably be a loop-until-empty in here somewhere
- A good pattern for ISRs is:
  - read pending ISRs
  - clear pending ISRs
  - process pending ISRs
This prevents race conditions when the ISR is re-triggered during processing (unless they are level sensitive or clearing has side effects)

vaishnavachath replied:
This is the last pending item, yes the mailbox is FIFO based and a loop until empty is needed to clear the FIFO, since most of the testing has been for signalling/single word transfer ,this was not caught , I am working on fixing this.
Ref: https://github.com/torvalds/linux/blob/master/drivers/mailbox/omap-mailbox.c#L206

gramsay0 commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+
+	irqstatus = regs->irq_regs[cfg->usr_id].enable_set;
+	irqstatus |= MAILBOX_IRQ_NEWMSG(channel);
+	regs->irq_regs[cfg->usr_id].enable_set = irqstatus;
```
Some of these functions should probably have a spinlock if they can be accessed from multiple threads

vaishnavachath replied:
Done , added

gramsay0 reviewed (commented):
It might be worth adding TI omap regex's to maintainers.yml as well.

Great work!

dnltz commented on boards/phytec/phycore_am62x/doc/index.rst:
```diff
@@ -40,6 +40,8 @@ features:
 | UART      | on-chip    | serial port                         |
 +-----------+------------+-------------------------------------+
+| Mailbox   | on-chip    | IPC Mailbox                         |
++-----------+------------+-------------------------------------+
```
@vaishnavachath Just figured out you added this only to the A53 core and not for the phyboard-lyra M4. I would appreciate if you could update the M4 documentation as well.

vaishnavachath replied:
Added for both Phyboard Lyra M4 and A53

dnltz commented:
@vaishnavachath can you please rebase?

cfriedt commented on dts/bindings/mbox/ti,omap-mailbox.yaml:
```diff
@@ -0,0 +1,23 @@
+
+compatible: "ti,omap-mailbox"
+
+include: base.yaml
```
```suggestion
include: [base.yaml, mailbox-controller.yaml]
```

vaishnavachath replied:
Done, Thank you for the review.

cfriedt commented on dts/bindings/mbox/ti,omap-mailbox.yaml:
```diff
@@ -0,0 +1,23 @@
+    required: true
+    description: User ID for processor
+
+mbox-cells:
+  - channel
```
This can be removed if you include `mailbox-controller.yaml`

vaishnavachath replied:
The default value is not mentioned in the `mailbox-controller.yaml` and thus it defaults to 0 instead of 1, please let know if we should set the default as 1 and then remove the redundant `mbox-cells` entry in most of the bindings.

cfriedt reviewed (commented):
Just some minor DT suggestions.

cfriedt commented:
@vaishnavachath - PTAL again

jan-kiszka commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,196 @@
+		DT_INST_REG_SIZE(idx), K_MEM_CACHE_NONE);	\
+		data->mapped_base = (void *)addr;	\
+		IRQ_CONNECT(DT_INST_IRQN(idx), DT_INST_IRQ(idx, priority), omap_mailbox_isr,	\
+			    DEVICE_DT_INST_GET(idx), 0);	\
```
You want to pass `DT_INST_IRQ(idx, flags)` here as flags, not 0. Hopefully, all interrupt controllers interested in that will use `flags` in their bindings, not `sense` (see `uart_ns16550.c`, sigh).

vaishnavachath committed:
MAINTAINERS: Add OMAP and DAVINCI drivers under TI

TI Platform drivers are named according to coresponding
names in Linux kernel, thus add the legacy drivers (DaVinci,
OMAP) also under TI K3 platforms.

Signed-off-by: Vaishnav Achath <vaishnav.a@ti.com>

cfriedt reviewed (dismissed):
LGTM

jan-kiszka commented on drivers/mbox/mbox_ti_omap.c:
```diff
@@ -0,0 +1,230 @@
+		IRQ_CONNECT(DT_INST_IRQN(idx), DT_INST_IRQ(idx, priority), omap_mailbox_isr,       \
+			    DEVICE_DT_INST_GET(idx),                                               \
+			    COND_CODE_1(DT_INST_IRQ_HAS_CELL(idx, flags),		\
+			    (DT_INST_IRQ(idx, flags)), (0)));       \
```
Indention of the trailing `\` is off. I would also indent the second line like this:
```
COND_CODE_1(DT_INST_IRQ_HAS_CELL(idx, flags),
            (DT_INST_IRQ(idx, flags)), (0))); 
```

kartben commented:
Ping @carlocaione

kartben commented:
@vaishnavachath this need to be rebased, please.

vaishnavachath commented:
@kartben @cfriedt sorry for the delay, I have fixed the conflicts
